gerrytools/geometry/dissolve.py

```diff
@@ -33,3 +37,144 @@ def dissolve(
+
+    level_gdfs = {
+        level: gdf
+        for gdf, level in zip((block_gdf, bg_gdf, tract_gdf, county_gdf),
+                              ('block', 'bg', 'tract', 'county'))
+        if gdf is not None
+    }
+
+    # Build nesting indices of Census geometries.
+    # Block GeoIDs are 15-16 characters long. The first 15 characters are
+    # numerals representing the state (2) + county (3) + tract (6) +
+    # block group (1) + block (3). There is an optional one-character block
+    # suffix. (See https://www.census.gov/programs-surveys/geography/
+    # guidance/geo-identifiers.html)
+
+    level_prefixes = {'county': 5, 'tract': 11, 'bg': 12, 'block': 16}
+    level_geoids_to_blocks = _group_by_level(
+        {str(b): b for b in block_gdf.index}, level_prefixes)
+    level_geoids_to_assignments = _group_by_level(
+        block_assignment, level_prefixes)
+
+    # Find the minimal set of geometries representing each assignment.
+    levels = tuple(
+        level
+        for level, _ in sorted(level_prefixes.items(), key=lambda kv: kv[1])
+        if level in level_gdfs
+    )  # sorted largest geometry (smallest prefix) -> smallest geometry
+    level_geoms = {
+        level: dict(gdf.geometry)
+        for level, gdf in level_gdfs.items()
+    }
+    geoms_by_assignment = defaultdict(list)
+    blocks_by_assignment = defaultdict(set)
+    for block, assignment in block_assignment.items():
+        blocks_by_assignment[assignment].add(block)
+    remaining_blocks_by_assignment = deepcopy(blocks_by_assignment)
+    level_counts = Counter()
+
+    for level in levels:
+        level_assignments = level_geoids_to_assignments[level]
+        for level_geoid, assignments in level_assignments.items():
+            if len(assignments) == 1:
+                # This unit *may* be wholly contained in a single district.
+                assignment, = tuple(assignments)
+                remaining_blocks = remaining_blocks_by_assignment[assignment]
+                level_blocks = level_geoids_to_blocks[level][level_geoid]
+                if level_blocks.issubset(remaining_blocks):
+                    # If we match, save the geometry and remove the
+                    # corresponding blocks from the assignment's
+                    # remaining blocks.
+                    level_geom = level_geoms[level][level_geoid]
+                    geoms_by_assignment[assignment].append(level_geom)
+                    remaining_blocks_by_assignment[assignment] -= level_blocks
+                    level_counts[level] += 1
+
+    # Dissolve geometries by assignment.
+    dissolved_gdf = GeoDataFrame([
+        {'label': assignment, 'geometry': unary_union(geometries)}
+        for assignment, geometries in geoms_by_assignment.items()
+    ]).sort_values(by='label').set_index('label')
+    dissolved_gdf.crs = block_gdf.crs
+
+    # Basic consistency check: district areas should approximately match.
+    block_geoms = level_geoms['block']
```

Comment:
This could reasonably be broken out into a separate function—I can see this consistency check being useful for custom dissolve operations beyond this one.